### PR TITLE
Propagate locals, even if they have unpropagatable assignments somewhere

### PR DESCRIPTION
**Pull request by oli-obk**
Second try for https://github.com/rust-lang/rust/pull/71946#discussion_r422967292

r? @wesleywiser 

cc @rust-lang/wg-mir-opt @RalfJung

### CODE REVIEW

File: src/test/mir-opt/const_prop/mutable_variable_aggregate_partial_read/rustc.main.ConstProp.diff [hunk 1]
```diff
@@ -29,17 +29,26 @@
                                            // + ty: i32
                                            // + val: Value(Scalar(0x00000063))
                                            // mir::Constant
-                                           // + span: $DIR/mutable_variable_aggregate_partial_read.rs:6:11: 6:13
+-                                          // + span: $DIR/mutable_variable_aggregate_partial_read.rs:6:11: 6:13
++                                          // + span: $DIR/mutable_variable_aggregate_partial_read.rs:6:5: 6:13
```
Comment (wesleywiser) [inline src/test/mir-opt/const_prop/mutable_variable_aggregate_partial_read/rustc.main.ConstProp.diff:33]: I should pull 060e437b776e238a492372fbb278fdc27fef1128 out of #71911 and get that merged so we can eliminate this kind of noise from the const prop tests.

**Review by wesleywiser — APPROVED** [review]

File: src/librustc_mir/transform/const_prop.rs [hunk 2]
```diff
@@ -893,6 +893,11 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
                     // ```
                     // FIXME: we overzealously erase the entire local, because that's easier to
                     // implement.
+                    trace!(
+                        "propagation into {:?} failed.
+                        Nuking the entire site from orbit, it's the only way to be sure",
```
Comment (wesleywiser) [inline src/librustc_mir/transform/const_prop.rs:898]: 👏

**Comment by wesleywiser** [comment]
@oli-obk I took the liberty of pushing a commit to your branch which blesses the new tests again since #72220 landed.

r=me if you're ok with that 🙂

**Comment by RalfJung** [comment]
LGTM. @bors r=wesleywiser

**Pull request merged**